access-functions.php

```diff
@@ -401,6 +401,45 @@ function ( TypeRegistry $type_registry ) use ( $type_name, $config ) {
 	);
 }
 
+/**
+ * Given a Type Name, this removes the type from the entire schema
+ *
+ * @param string $type_name The name of the Type to remove.
+ *
+ * @since @todo
+ */
+function deregister_graphql_type( string $type_name ) : void {
+	// Prevent the type from being registered to the scheme directly.
+	add_filter(
+		'graphql_excluded_types',
+		function ( $excluded_types ) use ( $type_name ) : array {
+			// Transform the type name, just in case a user provided graphql_single_name instead.
+			$type_name = ucfirst( $type_name );
```

Comment:
I think elsewhere in the registry we convert the strings to all lowercase to help avoid nuances of `typeName` vs `TypeName` vs `typename`. 

I'm thinking maybe this should do the same, instead of `ucfirst()`

Reply:
Types are added to the `type_loader` using the `format_key()` function: https://github.com/wp-graphql/wp-graphql/blob/6de28e35389891781b8107e0404fa8e0c415c844/src/Registry/TypeRegistry.php#L630

Which just maps to `strtolower()`: https://github.com/wp-graphql/wp-graphql/blob/6de28e35389891781b8107e0404fa8e0c415c844/src/Registry/TypeRegistry.php#L179-L181

Reply:
My understanding is lowercase patterns are used for _internal_ array keys. but we need to be concerned about what users are going to pass to [for the `graphql_excluded_types` filter](https://github.com/wp-graphql/wp-graphql/pull/2557/files#diff-fea80346fad1aa40b8c8872a4d3a8f73a343b677fe4e8410e1224fcfb502a1a5R1186).

If you want we can normalize it in TypeRegistry::get_excluded_types(), and then safely use lowercase across the PR. Your call 😎

Reply:
I think I'd lean toward normalizing to lowercase as I think it will help prevent any edge cases. 

Like dynamically generated names too. 

Like

```php
foreach ( $post_types as $post_type ) {
  deregister_graphql_type( 'RootQueryTo' . $post_type . 'Connection' );
 
}
```

it wouldn't de-register. But, if the strings were normalized to all lowercase, it _would_ deregister.

Reply:
They'd probably want to use the upcoming `deregister_graphql_connection()`  ;-)
```php
foreach( $post_type as $post_type ) {
  deregister_graphql_connection( 'RootQuery', $post_type, $maybe_an_options_array); 
  } 
```

More seriously (and possibly better suited for its own discussion/issue), I think _coding_ for `strtolower()` is a bad pattern.

We should be forgiving of user input of course, but all the types/fields except for CPT/Terms (aka the vast majority of them) use  `KebabCase` or `camelCase`. Abstracting them to `lowercase` doesnt just drop our ability to recreate the `snake_case` used programmatically in PHP, but also means WPGraphQL core needs to be coding defensively every single time type names _are used_ (hence all the `ucFirst()`s in this PR), instead of just on public method param types that user's might use in their own code.

I actually had turning the internal CPT/Tax names into `KebabCase` and moving as many of the transforms to the public api footprint instead of internal logic on my PR todo list ;-)

Reply:
(To clarify, I agree with the suggestion to use strtolower when comparing on top-level - going to go through the PR to hit those, just want to make sure the above delineation makes sense to you)

Reply:
> They'd probably want to use the upcoming deregister_graphql_connection() ;-)

lol. Ya, that was just _one_ example. 

This type of pattern exists in other things like ACF Flex Fields, likely Gutenberg Blocks when we go down that road, etc. 

For someone trying to extend the plugin, to have something like the middle word of a concatenated string _not_ do what they thought feels weird to me.

I think we want to encourage folks to use proper casing for Fields & Types when possible, but would hate for someone not adding `ucfirst()` in an example like above, cause their code to not work. It would be frustrating / difficult to debug, and probably even difficult to pinpoint when helping someone debug.

Reply:
Basically, in my mind, it makes sense to to the `strtolower()` at the deepest part of the codebase only. 

Like we do here: https://github.com/wp-graphql/wp-graphql/blob/6de28e35389891781b8107e0404fa8e0c415c844/src/Registry/TypeRegistry.php#L630

Only when the Type is formally being added to the registry is it normalized to all lowercase

I think this is ultimately where we should lowercase the excluded types and do what's needed to leave them out of the Type registry.